crypto/buffer.h

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright (c) 2023 Trezor Company s.r.o.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT HMAC_SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef __BUFFER_H
+#define __BUFFER_H
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+// Buffer reader struct, wraps a pointer to const data.
+typedef struct {
+  const uint8_t *data;
+  size_t size;
+  size_t pos;
+} BUFFER_READER;
+
+// Buffer writer struct, wraps a pointer to non-const data.
+// Always should have the same layout as BUFFER_READER. We consider it legal to
+// cast an BUFFER_WRITER pointer to an IOreader pointer. However, NOT the other
+// way around.
```

Comment:
I am not actually convinced that it makes sense to cast `BUFFER_WRITER` to `BUFFER_READER`.

Note to self: `IOreader` -> `BUFFER_READER`.

Reply:
the only reason for that in io.c was to reuse something like io_remaining which was called from everywhere. in this adaptation that is no longer the case, and you could do buffer_writer_remaining() and buffer_reader_remaining() separately either way